From e5312a9c481f91f2ca2cae1dc4176a78baaebf40 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Tue, 24 Jun 2014 10:00:15 +0100
Subject: [PATCH 22/24] Don't add rpaths to libraries.

---
 utils/config.mlp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/utils/config.mlp b/utils/config.mlp
index f758a9b483..ee17a737dc 100644
--- a/utils/config.mlp
+++ b/utils/config.mlp
@@ -54,8 +54,8 @@ let native_c_compiler =
   c_compiler ^ " " ^ ocamlopt_cflags ^ " " ^ ocamlopt_cppflags
 let native_c_libraries = "%%NATIVECCLIBS%%"
 let native_pack_linker = "%%PACKLD%%"
-let default_rpath = "%%RPATH%%"
-let mksharedlibrpath = "%%MKSHAREDLIBRPATH%%"
+let default_rpath = ""
+let mksharedlibrpath = ""
 let ar = "%%ARCMD%%"
 let supports_shared_libraries = %%SUPPORTS_SHARED_LIBRARIES%%
 let mkdll, mkexe, mkmaindll =
-- 
2.37.0.rc2

