diff --git a/Changes b/Changes
index a91fe1ecdb..6cf7e73033 100644
--- a/Changes
+++ b/Changes
@@ -118,6 +118,10 @@ OCaml 4.14.0 (28 March 2022)
   (Stephen Dolan, review by Xavier Leroy, Guillaume Munch-Maccagnoni,
    Jacques-Henri Jourdan, Damien Doligez and Leo White)
 
+- #12268: deliver `Out_of_memory` exception if domain creation fails
+  due to memory resource exhaustion.  It was previous always a `Failure`.
+  (Anil Madhavapeddy, review by David Allsopp)
+
 ### Code generation and optimizations (highlights):
 
 - #10595: Tail calls with up to 64 arguments are guaranteed to be compiled
